cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/matching.hpp

```diff
@@ -350,65 +356,70 @@ solution<Cpu> matching_engine<Cpu>::get_solution() {
     return std::move(engine_solutions);
 }
 
+template <typename Cpu>
+std::int64_t matching_engine<Cpu>::get_match_count() const {
+    return engine_solutions.get_solution_count();
+}
+
+template <typename Cpu>
+std::int64_t matching_engine<Cpu>::state_exploration() {
+    if (target->bit_representation) {
+        return state_exploration_bit();
+    }
+    else {
+        return state_exploration_list();
+    }
+}
+
 template <typename Cpu>
 void matching_engine<Cpu>::run_and_wait(global_stack<Cpu>& gstack,
                                         std::int64_t& busy_engine_count,
+                                        std::int64_t& current_match_count,
+                                        std::int64_t target_match_count,
                                         bool main_engine) {
     if (main_engine) {
         first_states_generator(hlocal_stack);
     }
     bool is_busy_engine = true;
     ONEDAL_ASSERT(pattern != nullptr);
-    if (target->bit_representation) { /* dense graph case */
-        for (;;) {
-            if (hlocal_stack.states_in_stack() > 0) {
-                while ((hlocal_stack.states_in_stack() > 5) && gstack.push(hlocal_stack))
-                    ;
-                ONEDAL_ASSERT(hlocal_stack.states_in_stack() > 0);
-                state_exploration_bit();
-            }
-            else {
-                gstack.pop(hlocal_stack);
-                if (hlocal_stack.empty()) {
-                    if (is_busy_engine) {
-                        is_busy_engine = false;
-                        dal::detail::atomic_decrement(busy_engine_count);
-                    }
-                    if (dal::detail::atomic_load(busy_engine_count) == 0)
-                        break;
+    for (;;) {
+        if (target_match_count > 0 &&
+            dal::detail::atomic_load(current_match_count) >= target_match_count) {
+            return;
+        }
+        if (hlocal_stack.states_in_stack() > 0) {
+            while ((hlocal_stack.states_in_stack() > 5) && gstack.push(hlocal_stack))
+                ;
+            ONEDAL_ASSERT(hlocal_stack.states_in_stack() > 0);
+            const std::int64_t prev_match_cout = get_match_count();
+            state_exploration();
+            if (target_match_count > 0) {
+                auto delta = get_match_count() - prev_match_cout;
+                if (delta > 0) {
+                    dal::detail::atomic_increment(current_match_count, delta);
                 }
-                else if (!is_busy_engine) {
-                    is_busy_engine = true;
-                    dal::detail::atomic_increment(busy_engine_count);
+                if (dal::detail::atomic_load(current_match_count) >= target_match_count) {
+                    return;
```

Comment:
You should also specify that engine is not busy for this case. Other threads can wait it infinite time otherwise.